content/2017/03/17/serializing-things.rst

```diff
@@ -0,0 +1,225 @@
+public: yes
+author: kit
+tags: [Celery, Django, Serialization, Code]
+image:
+  - src: '2017/serializing-things/celery-man.jpg'
+summary: |
+  Transmitting objects between web processes and worker processes is a
+  requirement of many modern web apps. Given that the safest way to do so is to
+  use a serialization format that only includes primitive data types, how can
+  we send custom objects around?
+
+
+Serializing Things
+==================
+
+Many Django apps need some kind of out-of-band processing. Perhaps you have to
+make calls to a separate API server, and don't want to nest the latency of HTTP
+calls. Or perhaps you need to perform a slow and complex post-processing step
+on user-submitted data. Whatever it is, you can't just put everything in your
+web processes. Some things demand worker processes consuming tasks from a
+queue.
+
+Typically we use Celery_ for this, though increasingly `Django Channels`_ is
+emerging as a viable option. In either case, there are some similar issues that
+arise from the use of a task queue broker; because of the indirection that an
+AMQP_ server or Redis_ or whatever introduces, you can't share memory, and
+therefore can't share in-memory objects directly. So, you somehow have to get
+data between the web requests and the worker processes. For now, we'll assume
+we're talking about Celery.
+
+If it's just primitive data, that's fine, you can serialize it. But what if
+it's not? What if you have a Django model that you want to pass to a task? This
+is hardly an uncommon need, but there are a few ways to approach the problem.
+
+First, of course, you can use the ``pickle`` Celery task serializer. This isn't
+terrible, but pickle does open you up to some `security concerns`_. You can
+mitigate them, but unless you feel confident that you know what you're doing,
+it's best to avoid exposing yourself to them in the first place. In the same
+way that you *can* sanitize all your database inputs and assemble DB queries
+through string concatenation – but would be well-advised to use prepared
+statements – you *can* ensure that everything that touches ``pickle`` is
+adequately sanitized.  I, for one, prefer to avoid using it, and save myself
+the worry.
```

Comment:
I'm not generally worried as long as I'm not transmitting pickles over the internet in plaintext and as long as I'm not pickling objects created by executing user input. But this is fair, I think.

Worth mentioning that using a different serialization format also makes it easier to interoperate with a worker running in a different language?